### PR TITLE
E2E Trade page

### PR DESCRIPTION
**Pull request by GrandSchtroumpf**

### CODE REVIEW

File: e2e/utils/trade.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  target: string;
+  sourceValue: string;
+  targetValue: string;
+}
+
+const getBalance = (page: Page, token: string) => {
+  return page.getByTestId(`balance-${token}`);
+};
+
+export const testTrade = async (page: Page, config: TradeConfig) => {
```
Comment (zavelevsky) [inline e2e/utils/trade.ts:18]: Here the assertion is mixed with the automation. The class should offer methods to selectPair, enter pay, etc. and assertion should be made by the caller by asking the class to fetch the values

File: e2e/utils/trade.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  // Token approval
+  const approvalModal = await waitModalOpen(page);
+  await approvalModal.getByTestId(`approve-${source}`).click();
+  const approvalMsg = await waitFor(page, `msg-${source}`);
+  await expect(approvalMsg).toHaveText('Approved');
+  await approvalModal.getByText('Confirm Trade').click();
+  await waitModalClose(page);
+
+  // Verify notification
```
Comment (zavelevsky) [inline e2e/utils/trade.ts:58]: assertion of notification should be elsewhere

File: e2e/utils/trade.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  await expect(notif.getByTestId('notif-title')).toHaveText('Success');
+  await expect(notif.getByTestId('notif-description')).toHaveText(
+    `Trading ${sourceValue} ${source} for ${target} was successfully completed.`
+  );
+
+  // Verify form empty
+  expect(form.getByLabel('You Pay')).toHaveValue('');
+  expect(form.getByLabel('You Receive')).toHaveValue('');
+
+  await navigateTo(page, '/debug');
```
Comment (zavelevsky) [inline e2e/utils/trade.ts:69]: This automation and assertion should be elsewhere

**Review by zavelevsky — CHANGES_REQUESTED** [review]

File: e2e/utils/notification.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  | 'create-strategy'
+  | 'renew-strategy'
+  | 'edit-strategy-name'
+  | 'withdraw-strategy'
+  | 'delete-strategy'
+  | 'change-rates-strategy'
+  | 'trade';
+
+export class NotificationDriver {
+  private notif = this.page.getByTestId(`notification-${this.type}`);
```
Comment (zavelevsky) [inline e2e/utils/notification.ts:17]: can this be called before the constructor??

Comment (GrandSchtroumpf) [reply e2e/utils/notification.ts:17]: No this is some typescript sugar, in this code will be compiled into the constructor

File: e2e/pages/strategy.spec.ts [hunk 5]
```diff
@@ -14,57 +17,46 @@ test.describe('Strategies', () => {
   });
 
   test('Create Limit Strategy ETH->DAI', async ({ page }) => {
-    test.setTimeout(120_000);
-    await waitFor(page, 'balance-DAI', 20_000);
+    test.setTimeout(180_000);
+    const config = {
```
Comment (zavelevsky) [inline e2e/pages/strategy.spec.ts:21]: absolute total nitpick, but if you put the config outside of the test then the test name can be dynamic and so there's only one place where you determine the tested tokens

File: e2e/pages/strategy.spec.ts [hunk 6]
```diff
@@ -13,71 +14,16 @@ test.describe('Strategies', () => {
...
-    await strategies.waitFor({ state: 'visible' });
-    await expect(strategies).toHaveCount(1);
-    const [strategy] = await strategies.all();
-    await expect(strategy.getByTestId('token-pair')).toHaveText('ETH/DAI');
-    await expect(strategy.getByTestId('status')).toHaveText('Active');
-    await expect(strategy.getByTestId('total-budget')).toHaveText('$3,344');
-    await expect(strategy.getByTestId('buy-budget')).toHaveText('10 DAI');
-    await expect(strategy.getByTestId('buy-budget-fiat')).toHaveText('$10.00');
-    await expect(strategy.getByTestId('sell-budget-fiat')).toHaveText('$3,334');
+  testCreateLimitStrategy({
```
Comment (zavelevsky) [inline e2e/pages/strategy.spec.ts:17]: I actually just meant:

```ts
const config =  {
  base: 'ETH',
  quote: 'DAI',
  buy: {
    price: '1500',
    budget: '10',
  },
  sell: {
    price: '1700',
    budget: '2',
  },
};
test(`Create Limit Strategy ${config.base}->${config.quote}`, async ({ page }) => {
...
```

I wouldn't put the test case into util - sorry if my earlier comment was confusing

File: e2e/pages/trade.spec.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import { navigateTo } from '../utils/operators';
+
+test.describe('Trade', () => {
+  test.beforeEach(async ({ page }) => {
+    await Promise.all([mockApi(page), setupImposter(page)]);
+  });
+
+  test('Buy ETH with USDC', async ({ page }) => {
+    test.setTimeout(120_000);
+    const config = {
```
Comment (zavelevsky) [inline e2e/pages/trade.spec.ts:15]: same here - move config out of the test case and change the test to

```ts
test(`${config.mode} ${config.target} with ${config.source}`, async ({ page }) => {
```

File: e2e/utils/debug.ts [hunk 8]
```diff
@@ -22,3 +22,11 @@ export const setupImposter = async (
     // Example: await waitFor(page, 'balance-DAI');
   }
 };
+
+export class DebugDriver {
```
Comment (zavelevsky) [inline e2e/utils/debug.ts:26]: file name should be DebugDriver.ts

File: e2e/utils/notification.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  | 'approve-error'
+  | 'create-strategy'
+  | 'renew-strategy'
+  | 'edit-strategy-name'
+  | 'withdraw-strategy'
+  | 'delete-strategy'
+  | 'change-rates-strategy'
+  | 'trade';
+
+export class NotificationDriver {
```
Comment (zavelevsky) [inline e2e/utils/notification.ts:16]: I'd say the file name should be NotificationDriver.ts

File: e2e/utils/strategy.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    await form.budget().fill(budget);
+    return form;
+  }
+  submit() {
+    return this.page.getByText('Create Strategy').click();
+  }
+}
+
+/** Generate a test that create a strategy */
+export const testCreateLimitStrategy = (config: LimitStrategyConfig) => {
```
Comment (zavelevsky) [inline e2e/utils/strategy.ts:66]: This belongs in the test file - see comment above

File: e2e/pages/strategy.spec.ts (hunk 6 above)
Comment (GrandSchtroumpf) [reply e2e/pages/strategy.spec.ts:17]: This works only if there is one test. Once I add more test into the describe I'll have to prefix the config or add a new scope. I would rather keep the config in the scope and hard code the name, as it was before, if it's ok with you.

Comment (zavelevsky) [reply e2e/pages/strategy.spec.ts:17]: Sure, but please take the test case out of utils

File: e2e/utils/strategy/MyStrategyDriver.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { Page } from 'playwright-core';
+
+export class MyStrategyDriver {
+  constructor(private page: Page) {}
+
+  firstStrategy() {
+    return this.page.getByTestId('first-strategy');
+  }
+
+  async getAllStrategy() {
```
Comment (zavelevsky) [inline e2e/utils/strategy/MyStrategyDriver.ts:10]: getAllStrategies :)

**Review by zavelevsky — APPROVED** [review]
LGTM!!!

**Pull request merged**